### PR TITLE
Fluid typography: pass theme.json settings to override merged theme data

### LINKED ISSUE #58135: Fluid font size application is dependent on the global state
### Description

I'm working with `WP_Theme_JSON` class. I feed it with a decode theme.json and use  to generate styles I want to use in a custom editor.
The issue is that even though I specify set `settings.typography.fluid` to `false` and provide values in pixels the styles generated by `WP_Theme_JSON::get_stylesheet` still contain the values converted to `clamp` function.

This seems to happen because `wp_get_typography_font_size_value` [checks the global settings](https://github.com/WordPress/gutenberg/blob/a3e4ee58c057cf2b91ba685b8ef3d8c3664ced44/lib/block-supports/typography.php#L458-L471) to see if the fluid font size is enabled. So if the site is configured to use the fluid font size it is always applied no matter what value I configure in data I pass to `WP_Theme_JSON`.

To add more context, my use case is a custom Gutenberg email editor, and I want to disable the fluid font size because `clamp` is not supported in many email clients.

### Step-by-step reproduction instructions

1) I have a WP site running on v 6.4, and I use The Twenty-Three theme.
2) I use the following code snippet
```php
$themeJsonData = '{
      "version": 2,
      "settings": {
        "typography": {
          "fluid": false
        }
      },
      "styles": {
        "elements": {
          "h1": {
            "typography": {
              "fontSize": "48px",
              "fluid": false
            }
          }
        }
      }
    }';
    $themeJson = new \WP_Theme_JSON(json_decode($themeJsonData, true), 'default');
    $css = $themeJson->get_stylesheet();
```  
3. The generated CSS contains`h1{font-size: clamp(27.894px, 1.743rem + ((1vw - 3.2px) * 2.285), 48px);}`
but I would expect `h1{font-size: 48px;}`

### Screenshots, screen recording, code snippet

_No response_

### Environment info

- WordPress 6.4
- I use the Gutenberg version distributed in the WordPress
- Twenty Twenty-Three theme

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ramonjd**
This is a proof of concept so far. Once an approach is settled, I'll:

- [x] as much as practicable, synchronize the JS equivalent (`getTypographyFontSizeValue`)
- [x] update tests and add more test coverage (for typography and layout.wideSize overrides)
- [x]  think about whether we need a follow up to handle preset generation via `value_func`
- [x] create a Core patch to sync https://github.com/WordPress/wordpress-develop/pull/6477

## What?
Working on something that resolves https://github.com/WordPress/gutenberg/issues/58135

Updates the second argument of `gutenberg_get_typography_font_size_value()` to be a theme_json settings array.

Add a backwards compatibility condition to handle booleans.

## Why?
To get fluid typography settings, `gutenberg_get_typography_font_size_value()` calls `gutenberg_get_global_settings()`.

`gutenberg_get_global_settings()` fetches merged theme.json data via `WP_Theme_JSON_Resolver_Gutenberg`.

However, to generate a stylesheet, theme.json settings are often passed directly to `WP_Theme_JSON_Gutenberg`.

Consider the following (modified) example provided by @costasovo in https://github.com/WordPress/gutenberg/issues/58135

```php
$theme_json_data = '{
    "version": 2,
    "settings": {
    "typography": {
        "fluid": false,
        "fontSizes" : [
                {
                    "name": "unos",
                    "size": "12px",
                    "slug": "unos"
                },
                {
                    "name": "dos",
                    "size": "20px",
                    "slug": "dos"
                }
            ]
        }
    },
    "styles": {
        "color": {
            "text": "#000000",
            "background": "#ffffff"
        },
            "typography": {
            "fontSize": "48px"
        },
        "elements": {
            "h1": {
            "typography": {
                "fontSize": "48px"
            }
            }
        },
        "blocks": {
            "test/blog": {
            "typography": {
                "fontSize": "48px"
            }
            }
        }
    }
}';
$my_theme_json = new \WP_Theme_JSON_Gutenberg( json_decode( $theme_json_data, true ), 'default' );
$css = $my_theme_json->get_stylesheet();
```

Because typography block supports looks at the global settings only, the font sizes and CSS vars will have a `clamp()` value in the returned stylesheet if the current theme has fluid typography activated. 

This is despite the explicit "false" in `$themeJsonData`. 

## How?

Passing theme.json settings to `gutenberg_get_typography_font_size_value()` as the second argument.

These settings will override any global theme settings.

Currently the argument expects a boolean, so I've added a backwards compat check.

## Testing Instructions
Using a theme that has activated fluid typography, add the above PHP snippet somewhere in the code (e.g., the theme's functions.php file) and log the output of `get_stylesheet()`.

Check that you can override the global theme settings.

Because this PR touches all fluid typography functionality, we'll also need to test basic output:

1. Check that fluid typography works as before (no regressions)
2. Check that the preset editor clamp values match the frontend output. Custom values are broken since https://github.com/WordPress/gutenberg/pull/56912#discussion_r1452946735

For example, I've been comparing themes against trunk:

Screenshot 2024-01-30 at 11 53 17 am

Also, run tests for glory!

JS
```
npm run test:unit packages/block-editor/src/components/global-styles/test/typography-utils.js
```

PHP
```
npm run test:unit:php:base -- --filter WP_Block_Supports_Typography_Test
```

### CODE REVIEW

File: lib/block-supports/typography.php [hunk 1]
```diff
@@ -438,11 +440,13 @@ function gutenberg_get_computed_fluid_typography_value( $args = array() ) {
...
  * }
- * @param bool  $should_use_fluid_typography An override to switch fluid typography "on". Can be used for unit testing. Default is `false`.
+ * @param bool|array $settings Optional Theme JSON settings array that overrides any global theme settings.
+ *                             As a bool, it acts as an override to switch fluid typography "on" (`true`) or "off" (`false`).
+ *                             Can be used for unit testing. Default is `array()`.
  *
  * @return string|null Font-size value or `null` if a size is not passed in $preset.
  */
-function gutenberg_get_typography_font_size_value( $preset, $should_use_fluid_typography = false ) {
+function gutenberg_get_typography_font_size_value( $preset, $settings = array() ) {
```
Comment (ramonjd) [inline lib/block-supports/typography.php:448]: 🤦🏻  

It should have been something like this in the first place.

I'm wondering if I shouldn't just change this to a general `$options` array to make adding further vars in the future easier?

So we'd be dealing with `$options['settings']`.

File: lib/class-wp-theme-json-gutenberg.php [hunk 2]
```diff
@@ -1154,7 +1154,7 @@ public function get_stylesheet( $types = array( 'variables', 'styles', 'presets'
 		if ( null === $origins ) {
 			$origins = static::VALID_ORIGINS;
 		}
-
+		//var_dump($this->get_settings());
```
Comment (ramonjd) [inline lib/class-wp-theme-json-gutenberg.php]: Delete this line

File: lib/block-supports/typography.php [hunk 3]
```diff
@@ -455,16 +459,28 @@ function gutenberg_get_typography_font_size_value( $preset, $should_use_fluid_ty
 		return $preset['size'];
 	}
 
+	/*
+	 * Backwards compatibility since 6.5.
+	 * If the settings argument is a boolean, it's a fluid typography override.
+	 */
+	if ( is_bool( $settings ) ) {
+		$settings = array(
+			'typography' => array(
```
Comment (ramonjd) [inline lib/block-supports/typography.php:468]: This bit of inelegant magic ensures backwards compatibility.

Searching https://wpdirectory.net/ I don't see anyone using the second bool argument for `wp_get_typography_font_size_value()`

File: lib/block-supports/typography.php [hunk 4]
```diff
@@ -455,16 +459,28 @@ function gutenberg_get_typography_font_size_value( $preset, $should_use_fluid_ty
...
-	$global_settings     = gutenberg_get_global_settings();
-	$typography_settings = isset( $global_settings['typography'] ) ? $global_settings['typography'] : array();
-	$layout_settings     = isset( $global_settings['layout'] ) ? $global_settings['layout'] : array();
-
-	$should_use_fluid_typography
-		= isset( $typography_settings['fluid'] ) &&
-		( true === $typography_settings['fluid'] || is_array( $typography_settings['fluid'] ) ) ?
-		true :
-		$should_use_fluid_typography;
+	$global_settings             = gutenberg_get_global_settings( array(), 'base' );
```
Comment (ramonjd) [inline lib/block-supports/typography.php]: This change is not needed. Left in by mistake.

File: lib/block-supports/typography.php (hunk 1 above)
Comment (ramonjd) [reply lib/block-supports/typography.php:448]: Actually, `$settings` makes sense I think as it harmonizes the PHP and JS function signatures.

$settings is an array already and can be extended.

File: lib/class-wp-theme-json-gutenberg.php [hunk 5]
```diff
@@ -2093,8 +2094,9 @@ protected static function compute_style_properties( $styles, $settings = array()
 				 * whether the incoming value can be converted to a fluid value.
 				 * Values that already have a clamp() function will not pass the test,
 				 * and therefore the original $value will be returned.
+				 * Pass the current theme_json settings to override any global settings.
 				 */
-				$value = gutenberg_get_typography_font_size_value( array( 'size' => $value ) );
+				$value = gutenberg_get_typography_font_size_value( array( 'size' => $value ), $settings );
```
Comment (ramonjd) [inline lib/class-wp-theme-json-gutenberg.php:2158]: Now, there's another consideration: what about preset generation?

Preset values are generated using this function too, however via a `value_func` callable property in `PRESETS_METADATA`. See:

https://github.com/WordPress/gutenberg/blob/b0162af871257438b8f994565016d7937ae010bd/lib/class-wp-theme-json-gutenberg.php#L160-L160

This callable only receives the font size value argument and not the settings:

https://github.com/WordPress/gutenberg/blob/b0162af871257438b8f994565016d7937ae010bd/lib/class-wp-theme-json-gutenberg.php#L1834-L1834

I'm not sure if it's kosher, but a follow up might have to solve this by sending `$settings` as the second arg.

```
$value      = call_user_func( $value_func, $preset, $settings );
```

File: lib/block-supports/typography.php [hunk 6]
```diff
@@ -430,19 +431,21 @@ function gutenberg_get_computed_fluid_typography_value( $args = array() ) {
  * @since 6.2.0 Added 'settings.typography.fluid.minFontSize' support.
  * @since 6.3.0 Using layout.wideSize as max viewport width, and logarithmic scale factor to calculate minimum font scale.
  * @since 6.4.0 Added configurable min and max viewport width values to the typography.fluid theme.json schema.
+ * @since 6.5.0 Deprecated bool argument $should_use_fluid_typography.
```
Comment (andrewserong) [inline lib/block-supports/typography.php]: Just double-checking, are we aiming to try to get this fix in for 6.5.0, or are we targeting 6.6 now?

File: lib/block-supports/typography.php [hunk 7]
```diff
@@ -455,22 +458,35 @@ function gutenberg_get_typography_font_size_value( $preset, $should_use_fluid_ty
...
-		true :
-		$should_use_fluid_typography;
+	// Fallback to global settings as default.
+	$global_settings             = gutenberg_get_global_settings();
+	$settings                    = wp_parse_args(
+		$settings,
+		$global_settings
+	);
+	$typography_settings         = isset( $settings['typography'] ) ? $settings['typography'] : array();
+	$should_use_fluid_typography = isset( $typography_settings['fluid'] ) && ! empty( $typography_settings['fluid'] );
```
Comment (andrewserong) [inline lib/block-supports/typography.php]: Tiny nit: do we still need the `isset()` half of this check since we have `! empty()` as the second half of it? I.e. would the following be equivalent?

```suggestion
	$should_use_fluid_typography = ! empty( $typography_settings['fluid'] );
```

File: phpunit/block-supports/typography-test.php [hunk 8]
```diff
@@ -316,291 +316,603 @@ public function test_gutenberg_get_typography_font_size_value( $font_size, $shou
...
 				),
-				'should_use_fluid_typography' => false,
-				'expected_output'             => null,
+				'settings'        => null,
+				'expected_output' => null,
 			),
 
-			'returns value when fluid is `false`'        => array(
-				'font_size'                   => array(
+			'returns value when fluid config is empty`'  => array(
```
Comment (andrewserong) [inline phpunit/block-supports/typography-test.php]: It looks like this test is still testing when the fluid value is false. Was that the intention? If so, just a quick suggestion to restore the previous label:

```suggestion
			'returns value when fluid is `false`'  => array(
```

File: lib/block-supports/typography.php (hunk 6 above)
Comment (ramonjd) [reply lib/block-supports/typography.php]: I reckon 6.6. I was just about to push an update that changes this. Cheers!

File: lib/class-wp-theme-json-gutenberg.php (hunk 5 above)
Comment (ramonjd) [reply lib/class-wp-theme-json-gutenberg.php:2158]: Done the above, with tests.

`Passing $settings to the callbacks defined in static::PRESETS_METADATA.`

Settings is an object, so in theory it can be extended. 

cc @oandregal for thoughts

File: phpunit/class-wp-theme-json-test.php [hunk 9]
```diff
@@ -1395,6 +1395,61 @@ public function test_get_stylesheet_custom_root_selector() {
...
+			$theme_json->get_stylesheet()
+		);
+		unregister_block_type( 'test/clamp-me' );
```
Comment (andrewserong) [inline phpunit/class-wp-theme-json-test.php]: Tiny nit: if the assertion fails, the `unregister_block_type` call will never be reached. Shall we move this call to just above the `$this->assertSame` line?

**Review by andrewserong — COMMENTED** [review]
This is testing great for me @ramonjd, and the code looks good!

I was mostly testing just before your last update, so I can give it a quick re-test with the presets included. But so far:

✅ Post editor output is as on `trunk`
✅ Site editor output is as on `trunk`
✅ Refactorings look good
✅ Output with provided PHP snippet is looking good, too

Nice work 👍

File: phpunit/class-wp-theme-json-test.php (hunk 9 above)
Comment (ramonjd) [reply phpunit/class-wp-theme-json-test.php]: No harm in that! 🌮

**Comment by ramonjd** [comment]
> I was mostly testing just before your last update, so I can give it a quick re-test with the presets included.

I'm grateful for the testing! Sorry you had to go through that before the presets update 🙏🏻 

I think that change is okay - seems like a useful addition (and fixes the output too!) but I pinged for a second opinion. Keen to hear your thoughts as well.

🍺

**Review by andrewserong — APPROVED** [review]
Looking good to me, thanks for the updates!

Custom PHP presets handled with `fluid`:

image

Custom PHP presets handled with `fluid` set to `false`:

image

> ...but I pinged for a second opinion. Keen to hear your thoughts as well.

Overall I really like the idea here, it makes a lot of sense to me to allow passing in of custom settings, and the way you've handled it for backcompat sounds reasonable to me, too 👍

Good call on getting a second opinion, but overall this LGTM! ✨

File: lib/class-wp-theme-json-gutenberg.php [hunk 10]
```diff
@@ -1824,7 +1825,7 @@ protected static function get_settings_values_by_slug( $settings, $preset_metada
 					is_callable( $preset_metadata['value_func'] )
 				) {
 					$value_func = $preset_metadata['value_func'];
-					$value      = call_user_func( $value_func, $preset );
+					$value      = call_user_func( $value_func, $preset, $settings );
```
Comment (ramonjd) [inline lib/class-wp-theme-json-gutenberg.php:1887]: Not sure yet, but this might also provide a path towards CSS var resolution for layout values. So, 


```json
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "var(--wp--custom--layout--content-size)",
			"wideSize": "1100px"
		},
		"typography": {
			"fluid": {
				"maxViewportWidth": {
					"ref": "settings.layout.contentSize"
				},
				"minViewportWidth": "600px"
			}
		}
	},
```

See:

- https://github.com/WordPress/gutenberg/issues/53525
- https://github.com/WordPress/gutenberg/issues/55070

**Comment by ramonjd** [comment]
Tentative ping to cc @WordPress/gutenberg-core if folks have any opinions to the [changes to class WP_Theme_JSON](https://github.com/WordPress/gutenberg/pull/58362/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R1828) in this PR:

**TL;DR** Passing theme.json `$settings` as a second argument to the `value_func` callbacks defined in `static::PRESETS_METADATA`.

See: 
https://github.com/WordPress/gutenberg/blob/671fbf961a7a091a3163ecf25e1a045973f458d1/lib/class-wp-theme-json-gutenberg.php#L120-L120

**Comment by ramonjd** [comment]
I'll go ahead and merge this one. If there's any contention about the approach then I'm happy revisit/revert.

**Comment by ramonjd** [comment]
Sorry, I forgot to add co-contribs 😢 

```
Unlinked contributors: costasovo.

Co-authored-by: ramonjd <ramonopoly@git.wordpress.org>
Co-authored-by: andrewserong <andrewserong@git.wordpress.org>
```

**Comment by ramonjd** [comment]
## Dev note for WordPress 6.6


Some theme.json presets require custom logic to generate their values, for example, when converting font size preset values to `clamp()` values.

The custom logic is handled by callback functions defined against the `value_func` key in `WP_Theme_JSON::PRESETS_METADATA`. The callback functions are invoked in `WP_Theme_JSON::get_settings_values_by_slug()`.

In WordPress 6.6, settings of the current `WP_Theme_JSON` instance, [are now passed to these callback functions](https://core.trac.wordpress.org/ticket/61118). The permits callback functions to return values that rely on other settings in the theme.json tree. 

In the case of font sizes presets, [it fixes a bug](https://github.com/WordPress/gutenberg/issues/58135) whereby the callback function — `wp_get_typography_font_size_value()` — was not taking into account settings values passed directly to the `WP_Theme_JSON` class.
